core/src/main/scala/org/bitcoins/core/psbt/PSBT.scala

```diff
@@ -0,0 +1,1870 @@
+      .asInstanceOf[Vector[key.RecordType]]
+  }
+}
+
+object PSBTMap {
+  final val separatorByte: Byte = 0x00.byteValue
+}
+
+sealed trait PSBTMapFactory[
+    RecordType <: PSBTRecord, MapType <: PSBTMap[RecordType]]
+    extends Factory[MapType] {
+  def recordFactory: Factory[RecordType]
+
+  def constructMap(elements: Vector[RecordType]): MapType
+
+  lazy val empty: MapType = constructMap(Vector.empty)
+
+  override def fromBytes(bytes: ByteVector): MapType = {
+    @tailrec
+    def loop(
+        remainingBytes: ByteVector,
+        accum: Vector[RecordType]): Vector[RecordType] = {
+      if (remainingBytes.head == PSBTMap.separatorByte) {
+        accum
+      } else {
+        val record = recordFactory.fromBytes(remainingBytes)
+        val next = remainingBytes.drop(record.bytes.size)
+
+        loop(next, accum :+ record)
+      }
+    }
+
+    constructMap(loop(bytes, Vector.empty))
+  }
+}
+
+case class GlobalPSBTMap(elements: Vector[GlobalPSBTRecord])
+    extends PSBTMap[GlobalPSBTRecord] {
+  import org.bitcoins.core.psbt.GlobalPSBTRecord._
+  import org.bitcoins.core.psbt.PSBTGlobalKeyId._
+  require(getRecords(UnsignedTransactionKeyId).nonEmpty,
+          "A GlobalPSBTMap must have a Unsigned Transaction")
+
+  def unsignedTransaction: UnsignedTransaction = {
+    getRecords(UnsignedTransactionKeyId).head
+  }
+
+  def extendedPublicKeys: Vector[XPubKey] = {
+    getRecords(XPubKeyKeyId)
+  }
+
+  def version: Version = {
+    getRecords(VersionKeyId).headOption.getOrElse(Version(UInt32.zero))
+  }
+
+  private def getRecords(key: PSBTGlobalKeyId): Vector[key.RecordType] = {
+    super.getRecords(key, PSBTGlobalKeyId)
+  }
+
+  /**
+    * Takes another GlobalPSBTMap and adds all records that are not contained in this GlobalPSBTMap
+    * @param other GlobalPSBTMap to be combined with
+    * @return A GlobalPSBTMap with the combined data of the two GlobalPSBTMaps
+    */
+  def combine(other: GlobalPSBTMap): GlobalPSBTMap = {
+    require(
+      this
+        .getRecords(UnsignedTransactionKeyId)
+        .head == other.getRecords(UnsignedTransactionKeyId).head,
+      "Cannot combine GlobalPSBTMaps with different unsigned transactions"
+    )
+    GlobalPSBTMap((this.elements ++ other.elements).distinct)
+  }
+}
+
+object GlobalPSBTMap extends PSBTMapFactory[GlobalPSBTRecord, GlobalPSBTMap] {
+
+  override def recordFactory: Factory[GlobalPSBTRecord] = GlobalPSBTRecord
+  override def constructMap(elements: Vector[GlobalPSBTRecord]): GlobalPSBTMap =
+    GlobalPSBTMap(elements)
+}
+
+case class InputPSBTMap(elements: Vector[InputPSBTRecord])
+    extends PSBTMap[InputPSBTRecord] {
+  import org.bitcoins.core.psbt.InputPSBTRecord._
+  import org.bitcoins.core.psbt.PSBTInputKeyId._
+
+  def nonWitnessOrUnknownUTXOOpt: Option[NonWitnessOrUnknownUTXO] = {
+    getRecords(NonWitnessUTXOKeyId).headOption
+  }
+
+  def witnessUTXOOpt: Option[WitnessUTXO] = {
+    getRecords(WitnessUTXOKeyId).headOption
```

Comment:
Why do we know that `.headOption` is right one? Couldn't it be at index 1 in the `Vector` returned here? The `getRecords()` stuff seems wonky

Reply:
For all the `Option[x]` ones there can only be one of them in the map